### PR TITLE
Fix Pacman regex for unmatched Arch package name

### PR DESCRIPTION
**Pull request by benizi**
##### SUMMARY

`ansible -m pacman -a upgrade=yes $(hostname)` failed due to not accounting for the `+` character in the `pacman -Qu` output line:

    libsigc++ 2.10.0-1 -> 2.10.1-1

[Per the Arch wiki][package-names], package names can contain alphanumeric characters, and any of {`@`, `.`, `_`, `+`, `-`}.

The existing `re` covered `_` (part of `\w` in Python), and `-` (explicitly included). This change adds `@`, `.`, and `+` (in ASCII-betical order).

[package-names]: https://wiki.archlinux.org/index.php/Arch_package_guidelines#Package_naming

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

pacman

##### ANSIBLE VERSION

Affected:

```
ansible 2.8.0.dev0 (devel 9a848ca883) last updated 2018/11/12 05:29:16 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/bhaskell/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/bhaskell/git/ansible/lib/ansible
  executable location = bin/ansible
  python version = 3.7.1 (default, Oct 22 2018, 10:41:28) [GCC 8.2.1 20180831]
```

Fixed:

```
ansible 2.8.0.dev0 (fix-pacman-upgrade-package-name-regex 82162c4ab6) last updated 2018/11/12 05:29:00 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/bhaskell/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/bhaskell/git/ansible/lib/ansible
  executable location = bin/ansible
  python version = 3.7.1 (default, Oct 22 2018, 10:41:28) [GCC 8.2.1 20180831]
```

##### ADDITIONAL INFORMATION

Before change (using `devel`, full version listed above):

```sh
$ archhost=bhaskell-asus
$ . hacking/env-setup >/dev/null 2>&1
$ bin/ansible -i $archhost, -m shell -a 'pacman -Qu | grep sigc' $archhost
bhaskell-asus | CHANGED | rc=0 >>
libsigc++ 2.10.0-1 -> 2.10.1-1

$ bin/ansible -i $archhost, -m pacman -a upgrade=yes -C $archhost 2>&1 |
… sed -e 1d -e '2i\{' |
… jq -r .msg
MODULE FAILURE
See stdout/stderr for the exact error
$
```

After change (using head of PR branch):

```sh
$ archhost=bhaskell-asus
$ . hacking/env-setup >/dev/null 2>&1
$ bin/ansible -i $archhost, -m shell -a 'pacman -Qu | grep sigc' $archhost
bhaskell-asus | CHANGED | rc=0 >>
libsigc++ 2.10.0-1 -> 2.10.1-1

$ bin/ansible -i $archhost, -m pacman -a upgrade=yes -C $archhost 2>&1 |
… sed -e 1d -e '2i\{' |
… jq '.packages|=(["..."]+map(select(contains("libsigc")))+["..."])'
{
  "changed": true,
  "msg": "618 package(s) would be upgraded",
  "packages": [
    "...",
    "libsigc++",
    "..."
  ]
}
$
```

### CODE REVIEW

File: lib/ansible/modules/packaging/os/pacman.py [hunk 1]
```diff
@@ -194,7 +194,7 @@ def upgrade(module, pacman_path):
     }
 
     if rc == 0:
-        regex = re.compile(r'([\w-]+) ((?:\S+)-(?:\S+)) -> ((?:\S+)-(?:\S+))')
+        regex = re.compile(r'([\w+\-.@]+) ((?:\S+)-(?:\S+)) -> ((?:\S+)-(?:\S+))')
```
Comment (tchernomax) [inline lib/ansible/modules/packaging/os/pacman.py]:
```suggestion
        # https://wiki.archlinux.org/index.php/Arch_package_guidelines#Package_naming
        # Package names can contain only alphanumeric characters and any of @, ., _, +, -.
        # All letters should be lowercase.
        regex = re.compile(r'([\w+-.@]+) ((?:\S+)-(?:\S+)) -> ((?:\S+)-(?:\S+))')
```

I think it should be useful to add some comment ; and to change `\-` to `-`  (why did you add a backslash before the hyphen ?)
What do you think ?

**Review by tchernomax — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/packaging/os/pacman.py (hunk 1 above)
Comment (benizi) [reply lib/ansible/modules/packaging/os/pacman.py]: > I think it should be useful to add some comment

Seems excessive to add the package naming guidelines directly in the code.  And the last item (`# All letters should be lowercase`) isn't reflected in the code.  (Nor is it reflected in package names -- there are packages on my system w/ capital letters.)

I agree that a comment makes sense, though. This is what I added:

```python
    if rc == 0:
        # Match lines of `pacman -Qu` output of the form:
        #   (package name) (before version-release) -> (after version-release)
        # e.g., "ansible 2.7.1-1 -> 2.7.2-1"
        regex = re.compile(r'([\w+\-.@]+) ((?:\S+)-(?:\S+)) -> ((?:\S+)-(?:\S+))')
```

> and to change `\-` to `-` (why did you add a backslash before the hyphen ?)

An unescaped hyphen (`-` without `\`) inside a set `[]` indicates a range, except when it's the first or last character in the set.

The old regex didn't need to have the hyphen escaped, because it was the last character. For the updated one (w/ `+`, `.`, and `@`):

`[\w+-.@]` (without the `\`) matches any character in:
- `\w` (roughly `r'[A-Za-z0-9_]'`)
- or `+-.` the range of characters `+` through `.`
- or `@`

`[\w+\-.@]` (with the `\`) matches any character in:
- `\w` (roughly `r'[A-Za-z0-9_]'`)
- or `+`
- or `\-` (only matches `-`; `\` just escapes its "range" semantics)
- or `.`
- or `@`

Comment (benizi) [reply lib/ansible/modules/packaging/os/pacman.py]: I also simplified the existing regex slightly by removing the unneeded non-capturing groups: `((?:\S+)-(?:\S+))` is exactly the same as `(\S+-\S+)`.

**Comment by benizi** [comment]
ready_for_review

File: lib/ansible/modules/packaging/os/pacman.py (hunk 1 above)
Comment (tchernomax) [reply lib/ansible/modules/packaging/os/pacman.py]: > Seems excessive to add the package naming guidelines directly in the code.

I personally like to have a link pointing to documentation ; but that's a personal taste.
I am totally ok with the comment you added.

> An unescaped hyphen (- without \) inside a set [] indicates a range, except when it's the first or last character in the set.
> The old regex didn't need to have the hyphen escaped, because it was the last character.

Very well explained, thanks !

**Review by tchernomax — APPROVED** [review]
LGTM

**Comment by afterburn** [comment]
Why am a I being mentioned by this anisbot?

**Comment by gundalow** [comment]
bot_status

**Comment by gundalow** [comment]
@afterburn You are listed as one of the authors https://github.com/ansible/ansible/blob/ee0fe2962acec66873b6faaac418ef56546d9822/lib/ansible/modules/packaging/os/pacman.py#L26

Is that incorrect?

**Comment by afterburn** [comment]
> @afterburn You are listed as one of the authors
> 
> [ansible/lib/ansible/modules/packaging/os/pacman.py](https://github.com/ansible/ansible/blob/ee0fe2962acec66873b6faaac418ef56546d9822/lib/ansible/modules/packaging/os/pacman.py#L26)
> 
> Line 26 in [ee0fe29](/ansible/ansible/commit/ee0fe2962acec66873b6faaac418ef56546d9822)
> 
>      - Afterburn 
> Is that incorrect?

Yes, most likely someone that owned this username before me contributed to this repo, not me. :)

**Comment by glitchcrab** [comment]
LGTM too.

**Comment by benizi** [comment]
Ping? Not fully familiar w/ Ansible's PR process, but I'm not sure whether there's anything I need to do to finish this out.

**Comment by tchernomax** [comment]
> Ping? Not fully familiar w/ Ansible's PR process, but I'm not sure whether there's anything I need to do to finish this out.

you can ping people asking them to "shipit" your PR.

* https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#when-will-your-pull-request-be-merged
* https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml

**Comment by benizi** [comment]
bot_status

**Comment by gundalow** [comment]
So we need some pacman users to review this

**Comment by glitchcrab** [comment]
OK, so I'm running this in containers as I don't particularly want to pollute my system.  I installed an outdated version of libsigc++:

```
[root@156c696f40b1 /]# pacman -Qu | grep libsig
libsigc++ 2.10.0-1 -> 2.10.1-1
```

Then ran an upgrade test which failed as expected (although I'm not sure of the exact error @benizi was seeing as the PR doesn't show that):
```
$ ansible -i inventory all -m pacman -a upgrade=yes --check

PLAY [Ansible Ad-Hoc] ************************************************************************************************************

TASK [pacman] ********************************************************************************************************************
Tuesday 19 March 2019  14:20:31 +0000 (0:00:00.054)       0:00:00.054 ********* 

fatal: [arch-container]: FAILED! => changed=false 
  ansible_facts:
    discovered_interpreter_python: /usr/bin/python
  module_stderr: |-
    <stdin>:18: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    Traceback (most recent call last):
      File "<stdin>", line 114, in <module>
      File "<stdin>", line 106, in _ansiballz_main
      File "<stdin>", line 49, in invoke_module
      File "/usr/lib/python3.7/imp.py", line 234, in load_module
        return load_source(name, filename, file)
      File "/usr/lib/python3.7/imp.py", line 169, in load_source
        module = _exec(spec, sys.modules[name])
      File "<frozen importlib._bootstrap>", line 630, in _exec
      File "<frozen importlib._bootstrap_external>", line 728, in exec_module
      File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
      File "/tmp/ansible_pacman_payload_zk88kvs8/__main__.py", line 482, in <module>
      File "/tmp/ansible_pacman_payload_zk88kvs8/__main__.py", line 453, in main
      File "/tmp/ansible_pacman_payload_zk88kvs8/__main__.py", line 236, in upgrade
    AttributeError: 'NoneType' object has no attribute 'group'
  module_stdout: ''
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
  rc: 1

PLAY RECAP ***********************************************************************************************************************
arch-container             : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```

I then checked out @benizi's PR, and ran the same:

```
$ ansible -i inventory all -m pacman -a upgrade=yes --check -v

PLAY [Ansible Ad-Hoc] ************************************************************************************************************

TASK [pacman] ********************************************************************************************************************
Tuesday 19 March 2019  14:26:19 +0000 (0:00:00.053)       0:00:00.053 ********* 
changed: [arch-container] => changed=true 
  msg: 1 package(s) would be upgraded
  packages:
  - libsigc++

PLAY RECAP ***********************************************************************************************************************
arch-container             : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
```

**Comment by pengwyn** [comment]
To help with the review, I can confirm that @benizi's PR also worked for me without any other problems.

**Comment by tchernomax** [comment]
@pengwyn if think this PR should be merged write "LGTM" or "shipit" in your comment (else it's not taken into account by ansibot).

**Comment by pengwyn** [comment]
Oh I see - thanks for letting me know. In that case:

LGTM

**Comment by indrajitr** [comment]
shipit

**Comment by indrajitr** [comment]
bot_status

**Review by indrajitr — APPROVED** [review]
